File: JavaApp/src/main/java/com/charlezz/javaapp/feature/local/UserAdapter.java
package com.charlezz.javaapp.feature.local;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.DiffUtil;

import com.charlezz.javaapp.BR;
import com.charlezz.javaapp.R;
import com.charlezz.javaapp.di.FragmentScope;
import com.charlezz.javaapp.feature.base.BindingPagedListAdapter;
import com.charlezz.javaapp.feature.base.ViewBindingHolder;

import javax.inject.Inject;

@FragmentScope
@SuppressWarnings("unchecked")
public class UserAdapter extends BindingPagedListAdapter {
    private static DiffUtil.ItemCallback<User> diffCallback = new DiffUtil.ItemCallback<User>() {
        @Override
        public boolean areItemsTheSame(@NonNull User oldItem, @NonNull User newItem) {
            return oldItem.getId() == newItem.getId();
        }

        @Override
        public boolean areContentsTheSame(@NonNull User oldItem, @NonNull User newItem) {
            return oldItem.getName().equals(newItem.getBirth()) &&
                    oldItem.getName().equals(newItem.getName());
        }
    };

    @Inject
    public UserAdapter() {
        super(diffCallback);
    }

    @NonNull
    @Override
    public ViewBindingHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        return new ViewBindingHolder(DataBindingUtil.inflate(LayoutInflater.from(viewGroup.getContext()), R.layout.view_user, viewGroup, false));
    }

    @NonNull
    @Override
    protected int getViewModelId(int viewType) {
        return BR.data;
    }

}

File: KotlinApp/src/main/java/com/charlezz/finalarchitecture/feature/local/UserAdapter.kt
package com.charlezz.finalarchitecture.feature.local

import androidx.annotation.NonNull
import androidx.recyclerview.widget.DiffUtil
import com.charlezz.finalarchitecture.BR
import com.charlezz.finalarchitecture.R
import com.charlezz.finalarchitecture.databinding.ViewUserBinding
import com.charlezz.finalarchitecture.feature.base.BaseAdapter
import com.charlezz.finalarchitecture.feature.base.BaseViewHolder
import javax.inject.Inject

class UserAdapter @Inject constructor(): BaseAdapter<User, ViewUserBinding, UserAdapter.PersonViewHolder>(BR.data,diffCallback) {

    override fun getViewHolderLayoutId(): Int = R.layout.view_user

    companion object {
        val diffCallback = object : DiffUtil.ItemCallback<User>() {
            override fun areContentsTheSame(@NonNull oldItem: User, @NonNull newItem: User): Boolean {
                return oldItem.id == newItem.id &&
                            oldItem.birth == newItem.birth &&
                            oldItem.name == newItem.name
            }

            override fun areItemsTheSame(oldItem: User, newItem: User): Boolean {
                return oldItem == newItem
            }
        }
    }
    inner class PersonViewHolder(binding: ViewUserBinding) : BaseViewHolder<ViewUserBinding>(binding)
}

